### PR TITLE
[PERF] base: avoid prefetching all fields when exporting

### PR DESCRIPTION
**Pull request by pivi-odoo**
## Description
When exporting a *wide* model, the process may face an out-of-memory error and be killed, even though that the large recordset is split into small batches of 1000 records via `splittor`. This is happening due to the field's prefetcher, on a model that is heavily extended, the reading of the first field to export, which happens at
https://github.com/odoo/odoo/blob/04a827e553cd6389a366c02d3a556de8c657ee22/odoo/models.py#L960 will read all fields (w/o those marked as `prefetch=False`) and put them into cache. On models that have some really large fields (like HTML descriptions), the limit can be quickly reached, even on the first split.

## Solution
Disable the field's prefetcher for the main loop during exporting. This way we avoid the OOM error, at the cost of doing a bit more queries (one foreach field we are exporting per batch of 1k records).

## Reference
opw-4202748

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by HydrionBurst** [comment]
What about making the number of split lines dynamic
`batch_size = 1000 * an_acceptable_field_number // len(model._fields)`

**Comment by pivi-odoo** [comment]
> > one for each field we are exporting per batch of 1k records
> 
> Surely it should give you pause that you're clearly creating an other performance issue?

@xmo-odoo between having an OOM that kills the request (blocking) and having a potentially slower export (non-blocking), the choice is quickly made.

Exporting is not a frequent operation, diff is small, risk is low.
But you will say: "They just have to export less at a time": in this db the problem is so bad, the OOM is triggered on the first batch of 1k, I can't tell them to export 100k records in batches of 1k manually. We recurrently have tickets with these kind of issues, we can't always do a manual interventions, the root cause needs to be addressed.

The ideal fix would be to explicitly `fetch` the list of fields we are exporting, but that's for a newer version (whenever the `fetch` was introduced). If you want I can retarget to a newer version version and use `fetch`.

I guess an alternative (for this version) would be to explicitly call `_read` for all fields we are exporting (essentially mimicking `fetch` in some sort), would you prefer such solution? 

> What about making the number of split lines dynamic
batch_size = 1000 * an_acceptable_field_number // len(model._fields)

@HydrionBurst well I doubt it would really solve the issue, here I have just 1 problematic field. :S 1k record seems already pretty fine to me as a small batch.

---
To recap, prefetching all fields during exporting, when a list of fields is explicitly passed doesn't make any sense (at least to me). Now *how* we go around addressing this, I will follow whatever propositions you prefer as codeowners, I just propose a `prefetch_fields=False` context key, if you have better proposition, let me know.

**Comment by pivi-odoo** [comment]
> I'm pretty sure requests are also killed if they take too long.

Yes, by default it's 15min for SaaS and Odoo.sh, but *if the prefetcher is disabled for exporting* the only variable left is time, which is something the end-user can control (by changing the number of records they are exporting).
If they face an OOM due to a fat field, even in a small batch of 1k, it's a blocking issue, requiring manual intervention from the support.

> How did you evaluate it exactly?

It's a one-liner, doesn't change logic/branching, there is little room for errors/oversights.
Pros: exports that were facing memory errors now pass (unless timing-out, but that's controllable)
Cons: exports that were barely passing under the 15min threshold, now can timeout (controllable by the end user by exporting less records at a time).

> Sounds to me like the real root cause is large description fields being prefetch. Because here you're working around the issue for imports, but you only need a client to have a server action working on a significant number of records to hit the exact same issue.

For this particular ticket, we know the issue, and we've always fixed it manually on each db 1 by 1, it's the fact that there are base64 inlined images in HTML `descriptions` instead of being attachment links.
The issue itself has been fixed in more recent versions, but it's not the case for 16.
We have scripts to migrate them, but it's a frequent issue, that I'm looking into a more permanent solution.
It's been already checked to set those problematic fields only as `prefetch=False` on their definition, but the idea was rejected.
As for Server Actions, we can't prevent people from shooting themselves in the foot, and it's not something that is being supported (if there is a ticket about such problem, we just point them to the problematic action, but we have no obligation "to make it work", it falls outside support warranty). But the same cannot be said about exporting.

> If you read the fields you care about, it should load only those fields (and their prefetch-able dependencies if computes) and put them in the cache, for regular access to retrieve them later.

I guess I could batch a `read` + prefetch_fields=False in advance (before looping). Sadly I'm kinda forced to disable the prefetcher for the `read`, because one of the fields for exporting could be a non-stored compute somwhere that reads on a value that isn't present as a dependency and basically re-triggers the prefetcher, voiding the idea of this patch.
Correct me if I'm wrong in my reflection.
